[sqllab] some frontend tests

mistercrunch opened the pull request.

mistercrunch committed:
[sqllab] some frontend tests

mistercrunch committed:
linting

ascott commented on caravel/assets/spec/javascripts/sqllab/common.js:
```diff
@@ -0,0 +1,229 @@
+export const alert = { bsStyle: 'danger', msg: 'Ooops' };
```
i think this file name would be more clear as `fixtures.js`

ascott commented on caravel/assets/spec/javascripts/sqllab/reducers_spec.js:
```diff
@@ -25,4 +26,73 @@ describe('sqlLabReducer', () => {
+    });
+    it('should remove a query editor', () => {
+      newState = r.sqlLabReducer(newState, actions.removeQueryEditor(qe));
+      expect(newState.queryEditors).to.have.lengthOf(1);
```
we could check the length here, or we could be more explicit and check that `newState.queryEditors` doesn't contain a `qe` with the same id, so we know the correct `qe` was removed.

ascott commented on caravel/assets/spec/javascripts/sqllab/reducers_spec.js:
```diff
@@ -25,4 +26,73 @@ describe('sqlLabReducer', () => {
+      expect(newState.tabHistory[newState.tabHistory.length - 1]).equals(defaultQueryEditor.id);
+    });
+    it('should not fail while setting attributes', () => {
+      newState = r.sqlLabReducer(state, actions.queryEditorSetDb(qe, 1));
```
tests that test one thing at a time are a good pattern that would be great to follow. it's easier/quicker to identify what is failing. so this could become 5 tests. we should also assert that the action didn't just not throw an error but changed the state in the correct way.
```
it('should set a query editor DB', () => {
  newState = r.sqlLabReducer(state, actions.queryEditorSetDb(qe, 1));
  // be explicit, add assertion
  // expect newState to have a qe object with the DB set to 1
});
```

ascott commented on caravel/assets/spec/javascripts/sqllab/reducers_spec.js:
```diff
@@ -25,4 +26,73 @@ describe('sqlLabReducer', () => {
+  describe('Tables', () => {
+    const state = Object.assign({}, initialState);
+    let newState;
+    it('should add a table', () => {
```
looks like this could be broken up into 4 tests for each action: `mergeTable`, `collapseTable`, `expandTable`, `removeTable`.

ascott commented on caravel/assets/spec/javascripts/sqllab/reducers_spec.js:
```diff
@@ -25,4 +26,73 @@ describe('sqlLabReducer', () => {
+  });
+  describe('Run Query', () => {
+    let newState = Object.assign({}, initialState);
+    it('should start', () => {
```
can we break this up into smaller, more specific tests plz.

ascott commented on caravel/assets/spec/javascripts/sqllab/VisualizeModal_spec.jsx:
```diff
@@ -0,0 +1,27 @@
+    show: true,
+    query: queries[0],
+  };
+  it('should just render', () => {
```
for naming tests, i prefer to leave out the should part on each test, so these would become `it('renders')`, `it('renders with props')`. when the tests run they'll then read like:
```
VisualizeModal
- renders ✔
- renders with props ✔
- renders a Modal ✔
```

ascott commented on caravel/assets/spec/javascripts/sqllab/reducers_spec.js:
```diff
@@ -2,17 +2,18 @@ import * as r from '../../../javascripts/SqlLab/reducers';
 import * as actions from '../../../javascripts/SqlLab/actions';
 import { describe, it } from 'mocha';
 import { expect } from 'chai';
+import { alert, table, initialState, query } from './common';
```
this would be more clear if the file was called `fixtures`

ascott commented on caravel/assets/spec/javascripts/sqllab/TabbedSqlEditors_spec.jsx:
```diff
@@ -0,0 +1,28 @@
+import { expect } from 'chai';
+import { initialState } from './common';
+
+describe('TabbedSqlEditors', () => {
```
it would be great to assert that the correct sub-components are found in this component. we should assert it renders a `<Tab />` and a `<SqlEditor />` for each tab. each tab should have a `<DropdownButton />` and each DropdownButton should have 3 `<MenuItem />` components. i see an opportunity to make the `TabbedSqlEditors` component smaller by pulling parts out into their own components, like a `Tabs` component and a`SQLEditorActionsButton` component.

ascott commented on caravel/assets/javascripts/SqlLab/components/QueryHistory.jsx:
```diff
@@ -7,6 +7,16 @@ import * as Actions from '../actions';
 
+const propTypes = {
+  queries: React.PropTypes.object,
+  tabHistory: React.PropTypes.array,
```
we should either make `tabHistory` and `actions` required, or add default props. 
`tabHistory: React.PropTypes.array.isRequired`,

ascott commented on caravel/assets/javascripts/SqlLab/components/QuerySearch.jsx:
```diff
@@ -7,6 +7,10 @@ import QueryTable from './QueryTable';
 import { STATUS_OPTIONS } from '../common';
 
+const propTypes = {
+  actions: React.PropTypes.object,
```
probably want to make `actions` required

ascott commented on caravel/assets/javascripts/SqlLab/components/VisualizeModal.jsx:
```diff
@@ -16,6 +12,16 @@ const CHART_TYPES = [
 
+const propTypes = {
+  onHide: React.PropTypes.func,
+  query: React.PropTypes.object,
```
is `query` required?

ascott commented on caravel/assets/spec/javascripts/sqllab/CopyQueryTabUrl_spec.jsx:
```diff
@@ -0,0 +1,23 @@
+      React.isValidElement(<CopyQueryTabUrl {...mockedProps} />)
+    ).to.equal(true);
+  });
+  it('shallow mounts', () => {
```
this test doesn't seem to add any additional value over the tests that test rendering above.

ascott replied:
we could add a test that asserts that a `CopyToClipboard` component is rendered...

ascott commented on caravel/assets/spec/javascripts/sqllab/HighlightedSql_spec.jsx:
```diff
@@ -0,0 +1,24 @@
+    expect(React.isValidElement(<HighlightedSql sql={sql} />))
+    .to.equal(true);
+  });
+  it('has a SyntaxHighlighter', () => {
```
this could be broken in to 2 smaller, more explicit tests.

ascott commented:
stoked to see tests for SQL Lab! i think we could make these test even more useful if we added some tests for component methods (`wrapper.instance().someMethod()`) and tested that the correct children are rendered as well for each component.

mistercrunch committed:
Addressing comments

ascott commented on caravel/assets/spec/helpers/browser.js:
```diff
@@ -17,4 +17,6 @@ Object.keys(document.defaultView).forEach((property) => {
 global.navigator = {
   userAgent: 'node.js',
+  platform: 'linux',
+  appName: 'Netscape',
```
cool, thanks for the explanation

mistercrunch commented:
Thank you for the review @ascott! I learned a lot through this PR and ended up refactoring things beyond tests, mostly around reducing the number of points of redux state injection.

I think I addressed all the comments.

mistercrunch commented:
Coverage for the whole repo:
```
Before
=============================== Coverage summary ===============================
Statements   : 12.7% ( 295/2322 )
Branches     : 2.23% ( 14/628 )
Functions    : 7.55% ( 40/530 )
Lines        : 12.75% ( 285/2236 )
================================================================================

After
=============================== Coverage summary ===============================
Statements   : 31.26% ( 703/2249 )
Branches     : 22.31% ( 139/623 )
Functions    : 21.87% ( 110/503 )
Lines        : 31.46% ( 683/2171 )
================================================================================
```
Details:
<img width="1280" alt="screen shot 2016-10-24 at 3 09 54 pm" src="https://cloud.githubusercontent.com/assets/487433/19665886/25f96f48-99fc-11e6-8981-96b08a4e4f78.png">

mistercrunch committed:
Addressing unaddressed comments

ascott commented on caravel/assets/javascripts/SqlLab/components/QueryHistory.jsx:
```diff
@@ -1,28 +1,26 @@
 
+const propTypes = {
+  queries: React.PropTypes.array.isRequired,
+  actions: React.PropTypes.object.isRequired,
```
if actions are required here, you don't need to include it in default props.

ascott commented on caravel/assets/javascripts/SqlLab/components/SqlEditor.jsx:
```diff
@@ -18,16 +18,29 @@ import 'brace/mode/sql';
+  networkOn: React.PropTypes.bool,
+  tables: React.PropTypes.array,
+  queries: React.PropTypes.array,
+  queryEditor: React.PropTypes.object,
```
is queryEditor required?

ascott commented on caravel/assets/javascripts/SqlLab/components/TabbedSqlEditors.jsx:
```diff
@@ -131,11 +136,18 @@ class TabbedSqlEditors extends React.Component {
-                latestQuery={latestQuery}
-                database={database}
-              />
+              {isSelected ?
```
this can be written more simply, we don't need the `condition ? then : else`ternary here.

```
{isSelected &&
  <SqlEditor 
    ...
  />
}
```

mistercrunch replied:
oh nice, I was looking for the perfect pattern for inline-if/then

ascott commented on caravel/assets/spec/javascripts/explore/components/QueryAndSaveBtns_spec.jsx:
```diff
@@ -1,13 +1,15 @@
+/* global beforeEach */
```
`beforeEach` can be imported from `mocha`

ascott commented on caravel/assets/spec/javascripts/sqllab/App_spec.jsx:
```diff
@@ -0,0 +1,18 @@
+  const mockedProps = {
+  };
+  it('renders', () => {
+    expect(React.isValidElement(<App />)).to.equal(true);
```
this is only testing that it's a valid element, it's not actually rendering, could update the test title to reflect this...

ascott commented on caravel/assets/spec/javascripts/sqllab/App_spec.jsx:
```diff
@@ -0,0 +1,18 @@
+
+
+describe('App', () => {
+  const mockedProps = {
```
unused var here.

ascott commented on caravel/assets/spec/javascripts/sqllab/Alerts_spec.jsx:
```diff
@@ -0,0 +1,26 @@
+  });
+  it('renders with props', () => {
+    expect(
+      React.isValidElement(<Alerts {...mockedProps}k>TEST</Alerts>)
```
looks like a typo, extra `k` after `{...mockedProps}`

mistercrunch commented on caravel/assets/javascripts/SqlLab/components/SqlEditor.jsx:
```diff
@@ -18,16 +18,27 @@ import 'brace/mode/sql';
 
 import SqlEditorLeftBar from './SqlEditorLeftBar';
 
+const propTypes = {
```
done

mistercrunch committed:
Touchups

mistercrunch commented:
@ascott I think I've ironed out all of your comments, your feedback was super valuable, thanks again!

ascott commented:
Generally looking good. Big diff! @mistercrunch, thanks for all the code coverage! 🚢
